Move `ast::Item::ident` into `ast::ItemKind`

nnethercote opened the pull request:
The follow-up to #138384, which did the same thing for `hir::ItemKind`.

r? @fmease

nnethercote committed:
Factor out some shared code.

`global_allocator_spans` and `alloc_error_handler_span` are identical
except for `name`.

nnethercote committed:
Remove useless `Option<Ident>` arg.

`FmtVisitor::visit_mac` has an `Option<Ident>` arg which is always
either `None` or `Some(kw::Empty)`, because `ItemKind::MacCall` always
has an empty ident. This value is passed through various functions until
it reaches `rewrite_macro_name`, which treats `None` and
`Some(kw::Empty)` the same.

In other words, the argument is useless. This commit removes it. There
is no change in behaviour. The commit also changes a few `symbol::Ident`
occurrences to `Ident` in `macros.rs`; `Symbol` is imported in that file
so `Ident` might as well be, too.

(This is a good example of why it's a bad idea for `Itemt` to have an
`ident` field when various item kinds don't have an identifier. It's
easy to get confused when "empty identifier" is used to mean "no
identifier". This will be fixed in a subsequent commit.)

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 87677af68745d3d7aa1c6539b3b98cd10c8bcfe8 with merge 960f7488ddd0a883d61daf2c4e954ec575bd58aa...
<!-- homu: {"type":"TryBuildStarted","head_sha":"87677af68745d3d7aa1c6539b3b98cd10c8bcfe8","merge_sha":"960f7488ddd0a883d61daf2c4e954ec575bd58aa"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13967814384/job/39109622904)
Build commit: 960f7488ddd0a883d61daf2c4e954ec575bd58aa (`960f7488ddd0a883d61daf2c4e954ec575bd58aa`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13967814384/job/39109622904"},"merge_sha":"960f7488ddd0a883d61daf2c4e954ec575bd58aa"} -->

rust-timer commented:
Queued 960f7488ddd0a883d61daf2c4e954ec575bd58aa with parent 4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579, future [comparison URL](https://perf.rust-lang.org/compare.html?start=4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579&end=960f7488ddd0a883d61daf2c4e954ec575bd58aa).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.5 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (960f7488ddd0a883d61daf2c4e954ec575bd58aa): [comparison URL](https://perf.rust-lang.org/compare.html?start=4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579&end=960f7488ddd0a883d61daf2c4e954ec575bd58aa&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579&end=960f7488ddd0a883d61daf2c4e954ec575bd58aa&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.3%, 0.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.3%, -0.2%] | 4     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.3% | [-0.3%, -0.2%] | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579&end=960f7488ddd0a883d61daf2c4e954ec575bd58aa&stat=max-rss)
<details>
<summary>Results (primary -0.7%, secondary -1.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1%  | [0.6%, 1.5%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 2.7%  | [2.7%, 2.7%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.6% | [-2.0%, -1.4%] | 4     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-2.8%, -1.8%] | 3     |
| All ❌✅ (primary)                 | -0.7% | [-2.0%, 1.5%]  | 6     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579&end=960f7488ddd0a883d61daf2c4e954ec575bd58aa&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579&end=960f7488ddd0a883d61daf2c4e954ec575bd58aa&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.0%, 0.1%]   | 33    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-1.0%, -0.0%] | 19    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.2% | [-1.0%, 0.1%]  | 52    |
</details>

**Bootstrap**: 773.468s -> 775.556s (0.27%)
**Artifact size**: 365.48 MiB -> 365.41 MiB (-0.02%)

nnethercote commented:
Performance is basically neutral. If anything, the improvements slightly outweigh the regressions, but all changes are very small.

@rustbot label: +perf-regression-triaged

nnethercote committed:
Simplify `ItemVisitorKind`.

Instead of putting the item inside it, just pass the ident and
visibility (the only things needed) alongside it where necessary.

This helps with the next commit, which will move the ident's location.
Specifically, it gets rid of the `match visitor_kind` in
`rewrite_type_alias`.

nnethercote committed:
Move `ast::Item::ident` into `ast::ItemKind`.

`ast::Item` has an `ident` field.

- It's always non-empty for these item kinds: `ExternCrate`, `Static`,
  `Const`, `Fn`, `Mod`, `TyAlias`, `Enum`, `Struct`, `Union`,
  `Trait`, `TraitAlias`, `MacroDef`, `Delegation`.

- It's always empty for these item kinds: `Use`, `ForeignMod`,
  `GlobalAsm`, `Impl`, `MacCall`, `DelegationMac`.

There is a similar story for `AssocItemKind` and `ForeignItemKind`.

Some sites that handle items check for an empty ident, some don't. This
is a very C-like way of doing things, but this is Rust, we have sum
types, we can do this properly and never forget to check for the
exceptional case and never YOLO possibly empty identifiers (or possibly
dummy spans) around and hope that things will work out.

The commit is large but it's mostly obvious plumbing work. Some notable
things.

- `ast::Item` got 8 bytes bigger. This could be avoided by boxing the
  fields within some of the `ast::ItemKind` variants (specifically:
  `Struct`, `Union`, `Enum`). I might do that in a follow-up; this
  commit is big enough already.

- For the visitors: `FnKind` no longer needs an `ident` field because
  the `Fn` within how has one.

- In the parser, the `ItemInfo` typedef is no longer needed. It was used
  in various places to return an `Ident` alongside an `ItemKind`, but
  now the `Ident` (if present) is within the `ItemKind`.

- In a few places I renamed identifier variables called `name` (or
  `foo_name`) as `ident` (or `foo_ident`), to better match the type, and
  because `name` is normally used for `Symbol`s. It's confusing to see
  something like `foo_name.name`.

nnethercote committed:
Ignore `#[test_case]` on anything other than `fn`/`const`/`static`.

`expand_test_case` looks for any item with a `#[test_case]` attribute
and adds a `test_path_symbol` attribute to it while also fiddling with
the item's ident's span.

This is pretty weird, because `#[test_case]` is only valid on
`fn`/`const`/`static` items, as far as I can tell. But you don't
currently get an error or warning if you use it on other kinds of items.

This commit changes things so that a `#[test_case]` item is modified
only if it is `fn`/`const`/`static`. This is relevant for moving idents
from `Item` to `ItemKind`, because some item kinds don't have an ident,
e.g. `impl` blocks.

The commit also does the following.
- Renames a local variable `test_id` as `test_ident`.
- Changes a `const` to `static` in
  `tests/ui/custom_test_frameworks/full.rs` to give the `static` case
  some test coverage.
- Adds a `struct` and `impl` to the same test to give some test coverage
  to the non-affected item kinds. These have a `FIXME` comment
  identifying the weirdness here. Hopefully this will be useful
  breadcrumbs for somebody else in the future.

nnethercote commented:
@fmease: thanks for reviewing, and apologies for the size of the final commit. I can move the four preliminary commits into a separate PR if that would be helpful.

bors commented:
:umbrella: The latest upstream changes (presumably #138923) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased.

bors commented:
:umbrella: The latest upstream changes (presumably #138933) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease commented on compiler/rustc_resolve/src/late/diagnostics.rs:
```diff
@@ -224,12 +224,17 @@ impl<'ast, 'ra: 'ast, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
                     && let Some(items) = self.diag_metadata.current_impl_items
                     && let Some(item) = items.iter().find(|i| {
-                        i.ident.name == item_str.name
+                        if let Some(ident) = i.kind.ident()
```
theoretically this could be made more terse:

```rs
i.kind.ident().is_some_and(|ident| ident.name == item_str.name && !sig.span.contains(item_span))
```

but I let you decide if you find that better / necessary

fmease reviewed (approved):
r=me after rebase and important nitpicks addressed. I like it very much, thanks!

I also posted a bunch of stylistic suggestions some of which probably aren't that important in retrospect. Feel free to ignore them (or address them) at your discretion. I don't want to require unnecessary work

fmease commented:
This PR is super bitrotty.
@bors p=1

nnethercote commented on compiler/rustc_builtin_macros/src/test.rs:
```diff
@@ -51,21 +51,26 @@ pub(crate) fn expand_test_case(
+            };
+            item.attrs.push(ecx.attr_name_value_str(sym::rustc_test_marker, test_path_symbol, sp));
+        }
+        _ => {}
```
This PR is already enormous, let's leave that for a follow-up.

nnethercote commented on tests/ui/custom_test_frameworks/full.rs:
```diff
@@ -25,4 +25,16 @@ impl example_runner::Testable for IsFoo {
-const TEST_2: IsFoo = IsFoo("foo");
+static TEST_2: IsFoo = IsFoo("foo");
+
+// FIXME: `test_case` is currently ignored on anything other than
+// fn/const/static. Should this be a warning/error?
```
I will update the comment to incorporate the extra information you gave.

nnethercote commented on compiler/rustc_builtin_macros/src/proc_macro_harness.rs:
```diff
@@ -214,15 +213,21 @@ impl<'a> Visitor<'a> for CollectProcMacros<'a> {
         };
 
-        if !is_fn {
+        // First up, make sure we're checking a bare function. If we're not then
+        // we're just not interested in this item.
+        //
+        // If we find one, try to locate a `#[proc_macro_derive]` attribute on it.
```
I will rephrase, it's not too hard.

nnethercote committed:
Address review comments.

nnethercote commented on compiler/rustc_resolve/src/build_reduced_graph.rs:
I'll change them. This is particularly good for `visit_assoc_item` -- the two matches on `item.kind` can be combined into one.

nnethercote commented on src/tools/clippy/clippy_lints/src/empty_line_after.rs:
```diff
@@ -375,21 +375,21 @@ impl EmptyLineAfter {
             kind: kind.descr(),
-            name: ident.name,
-            span: if span.contains(ident.span) {
+            name: if let Some(ident) = ident { ident.name } else { sym::dummy },
```
I'll add the FIXME and change it back to `sym::empty`. I prefer to preserve existing buggy behaviour rather than introducing new buggy behaviour :)

nnethercote commented on src/tools/clippy/clippy_utils/src/ast_utils/mod.rs:
```diff
@@ -518,6 +553,7 @@ pub fn eq_foreign_item_kind(l: &ForeignItemKind, r: &ForeignItemKind) -> bool {
+                && eq_id(*li, *ri)
                 && eq_generics(lg, rg)
                 && eq_opt_fn_contract(lc, rc)
                 && both(lb.as_ref(), rb.as_ref(), |l, r| eq_block(l, r))
```
Ugh, that's a bad place for a non-exhaustive match. I will make it exhaustive to avoid this error in the future.

nnethercote commented on src/tools/rustfmt/src/visitor.rs:
```diff
@@ -619,14 +623,15 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
+            AssocImplItem => visit::AssocCtxt::Impl,
             _ => unreachable!(),
         };
+        let skip_span = ai.span;
```
I'm now confused by the original code. `ai.span()` resolves to `Item::span`, but I can't find where that method is defined. I can see `Item::span_with_attributes`.

nnethercote replied:
Ok, I found it in `src/tools/rustfmt/src/spanned.rs`.

nnethercote replied:
I tried making `visit_assoc_item` match `visit_item`:
- use `ai.span` in the `skip_out_of_file_lines_range_visitor!` call
- use `ai.span()` in the `self.push_skipped_with_span` call

And I get a test failure:
```
Mismatch at tests/source/issue-4398.rs:6:
 
 impl Struct {
     /// Documentation for `foo`
-    #[rustfmt::skip] // comment on why use a skip here
+       #[rustfmt::skip] // comment on why use a skip here
     pub fn foo(&self) {}
 }
```
I agree that `AssocTraitItem` and `AssocImplItem` should be treated the same, so I will with `ai.span`.

nnethercote commented on compiler/rustc_ast/src/visit.rs:
Hmm, good catch. I will add `visit_ident` calls to `DefCollector` and `LateResolutionVisitor`. `UnsafeVisitor` doesn't need changing, given that it does almost nothing in the non-`walk_fn` case.

nnethercote commented:
Thanks for the detailed review. I made all the changes you suggested except where I mentioned otherwise. Should be good to go.

@bors r=fmease

bors commented:
:pushpin: Commit ec10833609aa63327437aabfaedfbe8a0edcc4d9 has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease ec10833609aa63327437aabfaedfbe8a0edcc4d9 -->
<!-- homu: {"type":"Approved","sha":"ec10833609aa63327437aabfaedfbe8a0edcc4d9","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit ec10833609aa63327437aabfaedfbe8a0edcc4d9 with merge ed201574c5d6117fb4a491db545c96fa4289ea9c...
<!-- homu: {"type":"BuildStarted","head_sha":"ec10833609aa63327437aabfaedfbe8a0edcc4d9","merge_sha":"ed201574c5d6117fb4a491db545c96fa4289ea9c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14189642664/job/39762827797)
Approved by: fmease
Pushing ed201574c5d6117fb4a491db545c96fa4289ea9c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"fmease","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14189642664/job/39762827797"},"merge_sha":"ed201574c5d6117fb4a491db545c96fa4289ea9c"} -->

rust-timer commented:
Finished benchmarking commit (ed201574c5d6117fb4a491db545c96fa4289ea9c): [comparison URL](https://perf.rust-lang.org/compare.html?start=97ea17b71aba0badfe9c8295f57a434bf79bdb72&end=ed201574c5d6117fb4a491db545c96fa4289ea9c&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=97ea17b71aba0badfe9c8295f57a434bf79bdb72&end=ed201574c5d6117fb4a491db545c96fa4289ea9c&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.3%, -0.1%] | 3     |
| Improvements ✅ <br /> (secondary) | -0.5% | [-0.8%, -0.3%] | 2     |
| All ❌✅ (primary)                 | -0.2% | [-0.3%, -0.1%] | 3     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=97ea17b71aba0badfe9c8295f57a434bf79bdb72&end=ed201574c5d6117fb4a491db545c96fa4289ea9c&stat=max-rss)
<details>
<summary>Results (primary -0.3%, secondary 2.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8%  | [0.4%, 1.5%]   | 4     |
| Regressions ❌ <br /> (secondary)  | 2.8%  | [1.2%, 3.9%]   | 7     |
| Improvements ✅ <br /> (primary)   | -1.8% | [-2.8%, -1.2%] | 3     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.1%, -2.1%] | 1     |
| All ❌✅ (primary)                 | -0.3% | [-2.8%, 1.5%]  | 7     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=97ea17b71aba0badfe9c8295f57a434bf79bdb72&end=ed201574c5d6117fb4a491db545c96fa4289ea9c&stat=cycles:u)
<details>
<summary>Results (secondary -3.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.2% | [-3.4%, -2.9%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=97ea17b71aba0badfe9c8295f57a434bf79bdb72&end=ed201574c5d6117fb4a491db545c96fa4289ea9c&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.0%, 0.1%]   | 34    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-1.0%, -0.0%] | 19    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.2% | [-1.0%, 0.1%]  | 53    |
</details>

**Bootstrap**: 773.194s -> 773.019s (-0.02%)
**Artifact size**: 366.00 MiB -> 365.92 MiB (-0.02%)
